### PR TITLE
Introspection template

### PR DESCRIPTION
**Pull request by kamilkisiela**
@dotansimha 

- [x] tests
- [x] write about it in README.md

`graphql-codegen-introspection-template`

### CODE REVIEW

File: packages/graphql-codegen-cli/src/codegen.ts [hunk 1]
```diff
@@ -40,7 +42,7 @@ export interface CLIOptions {
   skipDocuments?: any;
   config?: string;
   require?: string[];
-  overwrite: boolean;
+  overwrite?: boolean;
```
Comment (kamilkisiela) [inline packages/graphql-codegen-cli/src/codegen.ts]: TypeScript was showing a warning about this

File: packages/graphql-codegen-cli/src/codegen.ts [hunk 2]
```diff
@@ -29,6 +30,7 @@ process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
 export interface CLIOptions {
   file?: string;
   url?: string;
+  introspect?: boolean;
```
Comment (kamilkisiela) [inline packages/graphql-codegen-cli/src/codegen.ts]: there's a new flag that overwrites others

File: packages/graphql-codegen-cli/src/codegen.ts [hunk 3]
```diff
@@ -125,7 +129,7 @@ export const validateCliOptions = (options: CLIOptions) => {
     logger.warn(`--export is deprecated, use --schema instead.`);
   }
 
-  if (!template && !project) {
+  if (!template && !project && !introspect) {
```
Comment (kamilkisiela) [inline packages/graphql-codegen-cli/src/codegen.ts]: avoid throwing an error since we don't need `template` or `project` while dumping a schema

**Comment by kamilkisiela** [comment]
@dotansimha I added few comments

**Comment by kamilkisiela** [comment]
I assumed that `gql-gen` with `--introspect` overwrites other logic, like generating templates, files etc but I can make it work side by side in this PR or in near future.

**Comment by kamilkisiela** [comment]
@dotansimha I used template now

File: packages/templates/introspection/src/config.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import * as index from './template.handlebars';
+import { toJSON } from './helpers/to-json';
+
+export const config: GeneratorConfig = {
+  inputType: EInputType.SINGLE_FILE,
+  flattenTypes: true,
+  templates: {
+    index
+  },
+  primitives: {
```
Comment (kamilkisiela) [inline packages/templates/introspection/src/config.ts]: TypeScript says it's necessary to define them but I don't think this template even use it

File: packages/templates/introspection/src/config.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { EInputType, GeneratorConfig } from 'graphql-codegen-core';
+import * as index from './template.handlebars';
+import { toJSON } from './helpers/to-json';
+
+export const config: GeneratorConfig = {
+  inputType: EInputType.SINGLE_FILE,
+  flattenTypes: true,
```
Comment (kamilkisiela) [inline packages/templates/introspection/src/config.ts]: I don't know if I should skip that or leave it as is

File: packages/templates/introspection/src/config.ts (hunk 4 above)
Comment (dotansimha) [reply packages/templates/introspection/src/config.ts]: Primitives? it uses them in the template helpers

File: packages/templates/introspection/src/config.ts (hunk 5 above)
Comment (dotansimha) [reply packages/templates/introspection/src/config.ts]: Just leave it as it

**Comment by dotansimha** [comment]
Thanks @kamilkisiela !

**Pull request merged**